bpf/lib/ipip_termination.h

```diff
@@ -0,0 +1,78 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+
+
+#ifndef __LIB_IPIP_TERMINATION_H_
+#define __LIB_IPIP_TERMINATION_H_
+
+#ifdef ENABLE_IPIP_TERMINATION
+
+static __always_inline int
+decap_ipv4(struct __ctx_buff *ctx)
+{
+	struct iphdr *outer_ip4;
+	struct iphdr inner_ip4;
+	void *data, *data_end;
+
+	struct ipv4_ct_tuple tuple = {};
+	int ret, inner_ip4_off, outer_ip4_off = ETH_HLEN, outer_ip4_len, l4_off;
+
+	struct lb4_key key = {};
+	struct lb4_service *svc;
+
+	if (!revalidate_data_pull(ctx, &data, &data_end, &outer_ip4))
+		return DROP_INVALID;
+	if (outer_ip4->protocol != IPPROTO_IPIP) {
+		return CTX_ACT_OK;
+	}
+
+	outer_ip4_len = ipv4_hdrlen(outer_ip4);
+	inner_ip4_off = outer_ip4_off + outer_ip4_len;
+
+	if (ctx_load_bytes(ctx, inner_ip4_off, &inner_ip4, sizeof(inner_ip4)) < 0)
+		return DROP_INVALID;
+
+	ret = lb4_extract_tuple(ctx, &inner_ip4, inner_ip4_off, &l4_off, &tuple);
+	if (IS_ERR(ret)) {
+		/* Bypass */
+		return CTX_ACT_OK;
```

Comment:
In fact, there isn't a specific reason to bypass. It was defensive code added in case there were errors when this feature operated in our cloud. After testing in our cloud for a year, we've confirmed that this code works without issues. Additionally, since 'lb4_extract_tuple()` returns `DROP_CT_INVALID_HDR` in case of an error, it would be better to return an error instead of a bypass here as well. I will change this code along with other modifications.

Reply:
https://github.com/cilium/cilium/blob/27b50319b1a2767aeac8efef6d04d81588c46008/bpf/lib/ipip_termination.h#L33-L35